File: tools/message-intent-classifier/php/MessageClassifier/Classifier.php
<?php

declare(strict_types=1);

namespace MessageClassifier;

/**
 * Классификатор с семантикой engine_py/classifier.py:
 *
 * 1. normalizeText (mb_lower + ё->е + ASCII-whitespace fold);
 * 2. внутри плоскости правила отсортированы (priority asc, порядок загрузки);
 * 3. правило срабатывает, если любой pattern матчится И ни одна negation;
 * 4. победитель несёт ['category' => ..., 'reason' => 'keyword:<pattern>'];
 * 5. никто не сработал -> null для плоскости.
 */
final class Classifier
{
    public function __construct(private readonly Loader $loader) {}

    /** @return array<string, array{category: string, reason: string}|null> */
    public function classifyAll(?string $text): array
    {
        $normalized = Loader::normalizeText($text);
        $result = [];
        foreach (Loader::PLANES as $plane) {
            $result[$plane] = $this->classifyPlane($plane, $normalized);
        }

        return $result;
    }

    /** @return array{category: string, reason: string}|null */
    public function classifyPlane(string $plane, string $text): ?array
    {
        if ($text === '') {
            return null;
        }
        foreach ($this->loader->rulesFor($plane) as $rule) {
            if (! $rule['enabled']) {
                continue;
            }
            foreach ($rule['negations'] as $negation) {
                if (self::match((string) $negation, $text)) {
                    continue 2;
                }
            }
            foreach ($rule['patterns'] as $pattern) {
                if (self::match((string) $pattern, $text)) {
                    return [
                        'category' => (string) $rule['category'],
                        'reason' => 'keyword:'.$pattern,
                    ];
                }
            }
        }

        return null;
    }

    private static function match(string $pattern, string $subject): bool
    {
        // '/' и '#' запрещены в паттернах лоадером — разделитель безопасен.
        return preg_match('~'.$pattern.'~u', $subject) === 1;
    }
}

File: tools/message-intent-classifier/engine_py/classifier.py
"""Reference classifier: normalize, per-plane first-match-wins with negations.

Semantics (mirrored 1:1 in php/MessageClassifier/Classifier.php):

1. text -> lower + ё->е + whitespace fold;
2. per plane, rules sorted by (priority asc, load order);
3. a rule fires when any pattern matches AND no negation matches;
4. winner carries {"category": ..., "reason": "keyword:<pattern>"};
5. no rule fired -> plane value is None.
"""

from __future__ import annotations

import re

from .loader import PLANES, RuleSet, normalize_text


def classify(ruleset: RuleSet, text: str | None) -> dict[str, dict[str, str] | None]:
    normalized = normalize_text(text)
    result: dict[str, dict[str, str] | None] = {}
    for plane in PLANES:
        result[plane] = None
        if not normalized:
            continue
        for rule in ruleset.rules_by_plane.get(plane, []):
            if not rule.enabled:
                continue
            if any(neg.search(normalized) for neg in rule.negations):
                continue
            for pattern in rule.patterns:
                if isinstance(pattern.search(normalized), re.Match):
                    result[plane] = {
                        "category": rule.category,
                        "reason": f"keyword:{pattern.pattern}",
                    }
                    break
            if result[plane] is not None:
                break
    return result
